catalog/cataloger_merge.go

```diff
@@ -128,128 +165,99 @@ func hasCommitDifferences(tx db.Tx, leftID, rightID int64) (bool, error) {
 	return hasCommitDifferences, nil
 }
 
-func formatMergeMessage(leftBranch string, rightBranch string) string {
-	return fmt.Sprintf("Merge '%s' into '%s'", leftBranch, rightBranch)
-}
-
-func (c *cataloger) doMergeByRelation(ctx context.Context, tx db.Tx, relation RelationType, leftID, rightID int64, committer string, msg string, metadata Metadata) (CommitID, error) {
-	nextCommitID, err := getNextCommitID(tx)
-	if err != nil {
-		return 0, err
+func mergeBatch(tx db.Tx, mergeBatch mergeBatchType, previousMaxCommitID, nextCommitID CommitID, rightID int64) error {
+	// collect changes to apply  on the branch
+	paths := make([]string, 0, MergeBatchSize)
+	ctidArray := make([]string, 0, MergeBatchSize)
+	var tombstonePaths []string
+	for _, diffRec := range mergeBatch {
+		if diffRec.DiffType == DifferenceTypeRemoved || diffRec.DiffType == DifferenceTypeChanged {
+			paths = append(paths, diffRec.Entry.Path)
+		}
+		if (diffRec.DiffType == DifferenceTypeAdded || diffRec.DiffType == DifferenceTypeChanged) &&
+			diffRec.EntryCtid != nil {
+			ctidArray = append(ctidArray, *diffRec.EntryCtid)
+		}
+		if diffRec.DiffType == DifferenceTypeRemoved && diffRec.parentTombstoneNeeded {
+			tombstonePaths = append(tombstonePaths, diffRec.Entry.Path)
+		}
 	}
-
-	// do the merge based on the relation
-	previousMaxCommitID, err := getLastCommitIDByBranchID(tx, rightID)
-	if err != nil {
-		return 0, err
+	// apply changes
+	if len(paths) > 0 {
+		// set entries that exist in the right branch as deleted by entries that were removed or changed
+		setMaxCommit := sq.Update("catalog_entries").
+			Set("max_commit", previousMaxCommitID).
+			Where("branch_id = ? and max_commit = ?", rightID, MaxCommitID).
+			Where(sq.Eq{"path": paths})
+		sql, args, err := setMaxCommit.PlaceholderFormat(sq.Dollar).ToSql()
+		if err != nil {
+			return err
+		}
+		_, err = tx.Exec(sql, args...)
+		if err != nil {
+			return err
+		}
 	}
-	switch relation {
-	case RelationTypeFromParent:
-		err = c.mergeFromParent(ctx, tx, previousMaxCommitID, nextCommitID, leftID, rightID, committer, msg, metadata)
-	case RelationTypeFromChild:
-		err = c.mergeFromChild(ctx, tx, previousMaxCommitID, nextCommitID, leftID, rightID, committer, msg, metadata)
-	case RelationTypeNotDirect:
-		err = c.mergeNonDirect(ctx, tx, previousMaxCommitID, nextCommitID, leftID, rightID, committer, msg, metadata)
-	default:
-		return 0, ErrUnsupportedRelation
+	if len(ctidArray) > 0 {
+		// copy entries from left to right
+		internalSelect := sq.Select(int64Str(rightID), "path", "physical_address", "creation_date", "size", "checksum", "metadata", int64Str(int64(nextCommitID))).
```

Comment:
1. why do we need a static column with rightID if we have the value and we can use it in the outer query where part
2. nextCommitID got a type CommitID - you can add String() method to it so you can capture this one once.

Reply:
Done. Removed function int64Str